modules/currencies/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+		Ok(())
+	}
+
+	fn withdraw(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		Self::ensure_not_erc20(currency_id)?;
+		if amount.is_zero() {
+			return Ok(());
+		}
+		if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::withdraw(who, amount)?;
+		} else {
+			T::MultiCurrency::withdraw(currency_id, who, amount)?;
+		}
+		Self::deposit_event(RawEvent::Withdrawn(currency_id, who.clone(), amount));
+		Ok(())
+	}
+
+	fn can_slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> bool {
+		if currency_id.is_erc20() {
+			false
+		} else if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::can_slash(who, amount)
+		} else {
+			T::MultiCurrency::can_slash(currency_id, who, amount)
+		}
+	}
+
+	fn slash(currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) -> Self::Balance {
+		if currency_id.is_erc20() {
+			Default::default()
+		} else if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::slash(who, amount)
+		} else {
+			T::MultiCurrency::slash(currency_id, who, amount)
+		}
+	}
+}
+
+impl<T: Config> MultiCurrencyExtended<T::AccountId> for Module<T> {
+	type Amount = AmountOf<T>;
+
+	fn update_balance(currency_id: Self::CurrencyId, who: &T::AccountId, by_amount: Self::Amount) -> DispatchResult {
+		Self::ensure_not_erc20(currency_id)?;
+		if currency_id == T::GetNativeCurrencyId::get() {
+			T::NativeCurrency::update_balance(who, by_amount)?;
+		} else {
+			T::MultiCurrency::update_balance(currency_id, who, by_amount)?;
+		}
+		Self::deposit_event(RawEvent::BalanceUpdated(currency_id, who.clone(), by_amount));
+		Ok(())
+	}
+}
+
+impl<T: Config> MultiLockableCurrency<T::AccountId> for Module<T> {
+	type Moment = T::BlockNumber;
+
+	fn set_lock(lock_id: LockIdentifier, currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) {
+		if !currency_id.is_erc20() {
+			if currency_id == T::GetNativeCurrencyId::get() {
+				T::NativeCurrency::set_lock(lock_id, who, amount);
+			} else {
+				T::MultiCurrency::set_lock(lock_id, currency_id, who, amount);
+			}
+		}
+	}
+
+	fn extend_lock(lock_id: LockIdentifier, currency_id: Self::CurrencyId, who: &T::AccountId, amount: Self::Balance) {
+		if !currency_id.is_erc20() {
+			if currency_id == T::GetNativeCurrencyId::get() {
+				T::NativeCurrency::extend_lock(lock_id, who, amount);
+			} else {
+				T::MultiCurrency::extend_lock(lock_id, currency_id, who, amount);
+			}
+		}
+	}
+
+	fn remove_lock(lock_id: LockIdentifier, currency_id: Self::CurrencyId, who: &T::AccountId) {
+		if !currency_id.is_erc20() {
+			if currency_id == T::GetNativeCurrencyId::get() {
+				T::NativeCurrency::remove_lock(lock_id, who);
+			} else {
+				T::MultiCurrency::remove_lock(lock_id, currency_id, who);
+			}
+		}
+	}
+}
+
+impl<T: Config> MultiReservableCurrency<T::AccountId> for Module<T> {
```

Comment:
I think we should implement reserve for erc20 by transfer fund to a system account